agent/xds/delta_test.go

```diff
@@ -0,0 +1,760 @@
+		TypeUrl: EndpointType,
+		Nonce:   hexString(4),
+		RemovedResources: []string{
+			"fake-endpoints", // correcting the errant subscription
+		},
+	})
+
+	// And no other response yet
+	assertDeltaChanBlocked(t, envoy.deltaStream.sendCh)
+
+	// ACKs the listener
+	envoy.SendDeltaReqACK(t, ListenerType, 3, true, nil)
+
+	// ACK the endpoint removal
+	envoy.SendDeltaReqACK(t, EndpointType, 4, true, nil)
+
+	// If we re-subscribe to something even if there are no changes we get a
+	// fresh copy.
+	envoy.SendDeltaReq(t, EndpointType, &envoy_discovery_v3.DeltaDiscoveryRequest{
+		ResourceNamesSubscribe: []string{
+			"geo-cache.default.dc1.query.11111111-2222-3333-4444-555555555555.consul",
+		},
+	})
+
+	assertDeltaResponseSent(t, envoy.deltaStream.sendCh, &envoy_discovery_v3.DeltaDiscoveryResponse{
+		TypeUrl: EndpointType,
+		Nonce:   hexString(5),
+		Resources: makeTestResources(t,
+			makeTestEndpoints(t, snap, "tcp:geo-cache"),
+		),
+	})
+
+	envoy.SendDeltaReqACK(t, EndpointType, 5, true, nil)
+
+	// And no other response yet
+	assertDeltaChanBlocked(t, envoy.deltaStream.sendCh)
+
+	// TODO: test NACK
+
+	envoy.Close()
+	select {
+	case err := <-errCh:
+		require.NoError(t, err)
+	case <-time.After(50 * time.Millisecond):
+		t.Fatalf("timed out waiting for handler to finish")
+	}
+}
+
+func TestServer_DeltaAggregatedResources_v3_BasicProtocol_HTTP2(t *testing.T) {
+	aclResolve := func(id string) (acl.Authorizer, error) {
+		// Allow all
+		return acl.RootAuthorizer("manage"), nil
+	}
+	scenario := newTestServerDeltaScenario(t, aclResolve, "web-sidecar-proxy", "", 0)
+	mgr, errCh, envoy := scenario.mgr, scenario.errCh, scenario.envoy
+
+	sid := structs.NewServiceID("web-sidecar-proxy", nil)
+
+	// Register the proxy to create state needed to Watch() on
+	mgr.RegisterProxy(t, sid)
+
+	// Send initial cluster discover (empty payload)
+	// { "typeUrl": "type.googleapis.com/envoy.config.cluster.v3.Cluster" }
+	envoy.SendDeltaReq(t, ClusterType, nil)
+
+	// Check no response sent yet
+	assertDeltaChanBlocked(t, envoy.deltaStream.sendCh)
+
+	// Deliver a new snapshot (tcp with one http upstream)
+	snap := newTestSnapshot(t, nil, "http2", &structs.ServiceConfigEntry{
+		Kind:     structs.ServiceDefaults,
+		Name:     "db",
+		Protocol: "http2",
+	})
+	mgr.DeliverConfig(t, sid, snap)
+
+	runStep(t, "no-rds", func(t *testing.T) {
+		assertDeltaResponseSent(t, envoy.deltaStream.sendCh, &envoy_discovery_v3.DeltaDiscoveryResponse{
+			TypeUrl: ClusterType,
+			Nonce:   hexString(1),
+			Resources: makeTestResources(t,
+				makeTestCluster(t, snap, "tcp:local_app"),
+				makeTestCluster(t, snap, "http2:db"),
+				makeTestCluster(t, snap, "tcp:geo-cache"),
+			),
+		})
+
+		// Envoy then tries to discover endpoints for those clusters.
+		// {
+		//   "typeUrl": "type.googleapis.com/envoy.config.endpoint.v3.ClusterLoadAssignment",
+		//   "resourceNamesSubscribe": [
+		//     "api.default.dc1.internal.2902259a-31e7-62e0-fccc-0d482f347e98.consul"
```

Comment:
This SNI doesn't line up with the subscription below

Reply:
Removed the completely stale xDS json blobs from this.